be/src/storage/version_graph.cpp

```diff
@@ -239,66 +232,60 @@ void VersionGraph::construct_version_graph(const std::vector<RowsetMetaSharedPtr
-    int64_t end_vertex_index = _add_vertex_to_graph(version.second + 1);
-
-    // We assume this version is new version, so we just add two edges
-    // into version graph. add one edge from start_version to end_version
-    _version_graph[start_vertex_index].edges.emplace(end_vertex_index);
-
-    // We add reverse edge(from end_version to start_version) to graph
-    _version_graph[end_vertex_index].edges.emplace(start_vertex_index);
+    auto& start_vertex = _add_vertex_to_graph(version.first);
+    // Add version.second + 1 as new vertex of version graph if not exist.
+    auto& end_vertex = _add_vertex_to_graph(version.second + 1);
+
+    // Add one edge from start_version to end_version.
+    start_vertex->edges.emplace(version.second + 1);
+    // Add reverse edge from end_version to start_version.
+    end_vertex->edges.emplace(version.first);
 }
 
 Status VersionGraph::delete_version_from_graph(const Version& version) {
     int64_t start_vertex_value = version.first;
     int64_t end_vertex_value = version.second + 1;
-    auto start_iter = _vertex_index_map.find(start_vertex_value);
-    auto end_iter = _vertex_index_map.find(end_vertex_value);
+    auto start_iter = _version_graph.find(start_vertex_value);
+    auto end_iter = _version_graph.find(end_vertex_value);
 
-    if (start_iter == _vertex_index_map.end() || end_iter == _vertex_index_map.end()) {
+    if (start_iter == _version_graph.end() || end_iter == _version_graph.end()) {
         LOG(WARNING) << "vertex for version does not exists. "
                      << "version=" << version.first << "-" << version.second;
         return Status::NotFound("Not found version");
     }
 
-    int64_t start_vertex_index = start_iter->second;
-    int64_t end_vertex_index = end_iter->second;
     // Remove edge and its reverse edge.
     // When there are same versions in edges, just remove the frist version.
-    auto start_edges_iter = _version_graph[start_vertex_index].edges.find(end_vertex_value);
-    if (start_edges_iter != _version_graph[start_vertex_index].edges.end()) {
-        _version_graph[start_vertex_index].edges.erase(start_edges_iter);
+    auto start_edges_iter = start_iter->second->edges.find(end_vertex_value);
+    if (start_edges_iter != start_iter->second->edges.end()) {
+        start_iter->second->edges.erase(start_edges_iter);
     }
 
-    auto end_edges_iter = _version_graph[end_vertex_index].edges.find(start_vertex_value);
-    if (end_edges_iter != _version_graph[end_vertex_index].edges.end()) {
-        _version_graph[end_vertex_index].edges.erase(end_edges_iter);
+    auto end_edges_iter = end_iter->second->edges.find(start_vertex_value);
+    if (end_edges_iter != end_iter->second->edges.end()) {
+        end_iter->second->edges.erase(end_edges_iter);
     }
 
     return Status::OK();
 }
 
-int64_t VersionGraph::_add_vertex_to_graph(int64_t vertex_value) {
+std::shared_ptr<Vertex>& VersionGraph::_add_vertex_to_graph(int64_t vertex_value) {
```

Comment:
Is this version graph protected by lock? If so, I think std::unique_ptr is enough.
We should only use std::shared_ptr when they are shared by multi-threads. Otherwise I think we should use std::unique_ptr first.

Because std::shared_ptr will introduce atomic operation, and take more memory than std::unique_ptr

Reply:
use reference of shared_ptr will estimate the atomic operation?